### PR TITLE
asserts/repair.go: add "bases" and "modes" support to the repair assertion

### PR DESCRIPTION
**Pull request by anonymouse64**
The bases and modes fields will be used to filter what systems the repair
assertion should be run on, with the following constraints:

* A non-empty "modes" setting is only applicable for UC20+ systems that have
  recovery modes and invalid to be used with bases such as core or core18.
* An empty "bases" setting combined with a non-empty "modes" setting only
  applies to UC20+ systems.

Currently only specifically the "core20" and "core22" snaps are recognized as
boot base snaps that support UC20+, this may need to change in the future as we
have forks of the core20 snap for commercial projects, etc.

### CODE REVIEW

File: asserts/repair.go [hunk 1]
```diff
@@ -131,6 +153,38 @@ func assembleRepair(assert assertionBase) (Assertion, error) {
...
+	}
+
+	// verify that modes is a list of only "run", "recover" and "install"
+	if len(modes) != 0 {
+		for _, m := range modes {
+			// note we could import boot here to use i.e. boot.ModeRun, but that
+			// might make import cycles happen some day if boot needs to use
+			// the asserts package for some odd reason, so instead just use the
+			// values directly, they're unlikely to change now
+			if !strutil.ListContains([]string{"run", "recover", "install"}, m) {
```
Comment (pedronis) [inline asserts/repair.go:172]: There's always tension between letting people create nonsensical assertions vs preventing a old snapd to interpret a corner case for which we found a use case later but in this case whatever we do here needs to be supported n the repair services and runner. So I think we should match what we plan to do for now allow only "run" and "recover" here.

File: asserts/repair.go [hunk 2]
```diff
@@ -131,6 +153,38 @@ func assembleRepair(assert assertionBase) (Assertion, error) {
...
+				return nil, fmt.Errorf("header \"modes\" contains an invalid element: %q (valid values are run, recover and install)", m)
+			}
+		}
+
+		// if modes is non-empty, then bases must be core2X, i.e. core20+
+
+		// TODO: we should have a more centralized helper for identifying core2X
+		//       snaps to make it easier to add new base snaps
+		uc20AndUpBaseSnapRegexp := regexp.MustCompile(`^core2(0|2)$`)
```
Comment (pedronis) [inline asserts/repair.go:181]: this is probably too strict otoh. I think we should do the reverse and instead:

* always check that the content of bases are valid snap names with naming.ValidateSnap(name)
* only error here if bases contains "core" or "core18"

**Review by pedronis — COMMENTED** [review]
some suggestions

File: asserts/repair.go (hunk 1 above)
Comment (anonymouse64) [reply asserts/repair.go]: done, now install is invalid and we unit test this

File: asserts/repair.go (hunk 2 above)
Comment (anonymouse64) [reply asserts/repair.go]: ok, I changed it slightly to also error if the base snap name is `core16` too and independently verify all elements of `bases` are valid snap names

File: asserts/repair.go [hunk 3]
```diff
@@ -162,25 +162,33 @@ func assembleRepair(assert assertionBase) (Assertion, error) {
 		return nil, err
 	}
 
-	// verify that modes is a list of only "run", "recover" and "install"
+	// validate that all base snap names are valid snap names
+	for _, b := range bases {
+		if err := naming.ValidateSnap(b); err != nil {
+			return nil, fmt.Errorf("cannot use %q as element in \"bases\": %v", b, err)
```
Comment (pedronis) [inline asserts/repair.go:168]: invalid snap name %q in "bases" perhaps ?

File: asserts/repair.go [hunk 4]
```diff
@@ -162,25 +162,33 @@ func assembleRepair(assert assertionBase) (Assertion, error) {
...
+		}
+	}
+
+	// verify that modes is a list of only "run" and "recover"
 	if len(modes) != 0 {
 		for _, m := range modes {
 			// note we could import boot here to use i.e. boot.ModeRun, but that
```
Comment (pedronis) [inline asserts/repair.go:175]: we definitely don't want to import boot from asserts :)  (asserts is definitely used outside of snapd so its deps should stay lean and minimal and not too entangled with other parts)

File: asserts/repair.go [hunk 5]
```diff
@@ -162,25 +162,33 @@ func assembleRepair(assert assertionBase) (Assertion, error) {
...
+		// are _not_ UC20: core and core18
 
-		// TODO: we should have a more centralized helper for identifying core2X
-		//       snaps to make it easier to add new base snaps
-		uc20AndUpBaseSnapRegexp := regexp.MustCompile(`^core2(0|2)$`)
 		for _, b := range bases {
-			if !uc20AndUpBaseSnapRegexp.MatchString(b) {
+			// fail on uc16 and uc18 base snaps
+			if b == "core" || b == "core18" || b == "core16" {
 				return nil, fmt.Errorf("in the presence of a non-empty \"modes\" header, \"bases\" must only contain base snaps supporting Ubuntu Core 20 boot bases")
```
Comment (pedronis) [inline asserts/repair.go:192]: s/Ubuntu Core 20 boot bases/recovery modes ?

**Review by pedronis — COMMENTED** [review]
thank you, some nitpicks/suggestions about mainly the errors

**Review by pedronis — APPROVED** [review]

File: asserts/repair.go (hunk 3 above)
Comment (anonymouse64) [reply asserts/repair.go:168]: mmm that results in this for at least our current error unit test case:

```
assertion repair: invalid snap name \"foo....bar\" in \"bases\": invalid snap name: \"foo....bar\"
```

but maybe that's fine

**Review by mvo5 — APPROVED** [review]
Thank you

**Pull request merged**